File: app/src/main/java/com/example/facecheck/ui/auth/LoginActivity.java
package com.example.facecheck.ui.auth;

import android.content.Intent;
import android.os.Bundle;
import android.text.TextUtils;
import android.view.View;
import android.widget.Button;
import android.widget.CheckBox;
import android.widget.EditText;
import android.widget.ProgressBar;
import android.widget.TextView;
import android.widget.Toast;

import androidx.appcompat.app.AppCompatActivity;
import androidx.lifecycle.ViewModelProvider;

import com.example.facecheck.R;
import com.example.facecheck.MainActivity;
import android.content.SharedPreferences;
import com.airbnb.lottie.LottieAnimationView;
import android.animation.Animator;
import android.animation.AnimatorListenerAdapter;

public class LoginActivity extends AppCompatActivity {

    private EditText emailEditText, passwordEditText;
    private Button loginButton;
    private TextView registerTextView;
    private ProgressBar progressBar;
    private LoginViewModel loginViewModel;
    private CheckBox rememberPasswordCheckBox;
    private LottieAnimationView lottieLoginView;
    private View lottieOverlayLogin;
    private boolean navigated = false;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_login);

        // 初始化ViewModel
        loginViewModel = new ViewModelProvider(this).get(LoginViewModel.class);

        // 初始化视图
        emailEditText = findViewById(R.id.emailEditText);
        passwordEditText = findViewById(R.id.passwordEditText);
        loginButton = findViewById(R.id.loginButton);
        registerTextView = findViewById(R.id.registerTextView);
        progressBar = findViewById(R.id.progressBar);
        rememberPasswordCheckBox = findViewById(R.id.rememberPasswordCheckBox);
        lottieLoginView = findViewById(R.id.lottieLoginView);
        lottieOverlayLogin = findViewById(R.id.lottieOverlayLogin);

        // 观察ViewModel状态
        loginViewModel.uiState.observe(this, this::handleUiState);

        // 设置登录按钮点击事件
        loginButton.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                loginUser();
            }
        });

        // 设置注册文本点击事件
        registerTextView.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                // 跳转到注册页面
                Intent intent = new Intent(LoginActivity.this, RegisterActivity.class);
                startActivity(intent);
            }
        });

        SharedPreferences prefs = getSharedPreferences("user_prefs", MODE_PRIVATE);
        String role = prefs.getString("user_role", "");
        long teacherId = prefs.getLong("teacher_id", -1);
        long studentId = prefs.getLong("student_id", -1);
        if (("teacher".equals(role) && teacherId != -1) || ("student".equals(role) && studentId != -1)) {
            navigated = true;
            Intent intent = new Intent(LoginActivity.this, MainActivity.class);
            intent.addFlags(Intent.FLAG_ACTIVITY_CLEAR_TOP | Intent.FLAG_ACTIVITY_NEW_TASK);
            startActivity(intent);
            finish();
        }
    }
    
    private void handleUiState(LoginUiState state) {
        if (state instanceof LoginUiState.Initial) {
            progressBar.setVisibility(View.GONE);
            loginButton.setEnabled(true);
        } else if (state instanceof LoginUiState.Loading) {
            progressBar.setVisibility(View.VISIBLE);
            loginButton.setEnabled(false);
        } else if (state instanceof LoginUiState.Success) {
            LoginUiState.Success success = (LoginUiState.Success) state;
            progressBar.setVisibility(View.GONE);
            Toast.makeText(LoginActivity.this, "登录成功！", Toast.LENGTH_SHORT).show();
            
            SharedPreferences prefs = getSharedPreferences("user_prefs", MODE_PRIVATE);
            SharedPreferences.Editor ed = prefs.edit();
            ed.putString("user_role", success.role);
            if ("teacher".equals(success.role)) {
                ed.putLong("teacher_id", success.userId).remove("student_id");
            } else {
                ed.putLong("student_id", success.userId).remove("teacher_id");
                // 学生默认进入首页，避免加载教师专用页面
                getSharedPreferences("ui_prefs", MODE_PRIVATE).edit()
                        .putInt("nav_selected_id", R.id.nav_home).apply();
            }
            ed.apply();
            
            // 记住密码逻辑
            if (rememberPasswordCheckBox.isChecked()) {
                prefs.edit()
                    .putBoolean("remember_password", true)
                    .putString("saved_username", emailEditText.getText().toString().trim())
                    .putString("saved_password", passwordEditText.getText().toString().trim())
                    .apply();
            } else {
                prefs.edit()
                    .putBoolean("remember_password", false)
                    .remove("saved_username")
                    .remove("saved_password")
                    .apply();
            }
            // 显示登录成功 Lottie（telegram），执行1次后退出动画再跳转
            if (lottieLoginView != null && lottieOverlayLogin != null) {
                loginButton.setEnabled(false);
                lottieOverlayLogin.setVisibility(View.VISIBLE);
                lottieLoginView.setAnimation("lottie/telegram.json");
                lottieLoginView.setRepeatCount(0);
                lottieLoginView.addAnimatorListener(new AnimatorListenerAdapter() {
                    @Override
                    public void onAnimationEnd(Animator animation) {
                        lottieOverlayLogin.setVisibility(View.GONE);
                if (!navigated) {
                    navigated = true;
                    Intent intent = new Intent(LoginActivity.this, MainActivity.class);
                    intent.addFlags(Intent.FLAG_ACTIVITY_CLEAR_TOP | Intent.FLAG_ACTIVITY_NEW_TASK);
                    startActivity(intent);
                    finish();
                }
                    }
                });
                lottieLoginView.playAnimation();
            } else {
                // 回退：无动画视图则直接跳转
                if (!navigated) {
                    navigated = true;
                    Intent intent = new Intent(LoginActivity.this, MainActivity.class);
                    intent.addFlags(Intent.FLAG_ACTIVITY_CLEAR_TOP | Intent.FLAG_ACTIVITY_NEW_TASK);
                    startActivity(intent);
                    finish();
                }
            }
        } else if (state instanceof LoginUiState.Error) {
            LoginUiState.Error error = (LoginUiState.Error) state;
            progressBar.setVisibility(View.GONE);
            loginButton.setEnabled(true);
            Toast.makeText(LoginActivity.this, error.message, Toast.LENGTH_SHORT).show();
        }
    }

    private void loginUser() {
        String username = emailEditText.getText().toString().trim();
        String password = passwordEditText.getText().toString().trim();

        // 验证输入
        if (TextUtils.isEmpty(username)) {
            emailEditText.setError("请输入用户名");
            return;
        }

        if (TextUtils.isEmpty(password)) {
            passwordEditText.setError("请输入密码");
            return;
        }

        // 自动填充记住的用户名和密码
        SharedPreferences prefs = getSharedPreferences("user_prefs", MODE_PRIVATE);
        boolean rememberPassword = prefs.getBoolean("remember_password", false);
        if (rememberPassword) {
            String savedUsername = prefs.getString("saved_username", "");
            String savedPassword = prefs.getString("saved_password", "");
            emailEditText.setText(savedUsername);
            passwordEditText.setText(savedPassword);
            rememberPasswordCheckBox.setChecked(true);
        }

        // 调用ViewModel进行登录
        loginViewModel.login(username, password);
    }
    
    // 移除 onStart 自动跳转，避免多次回调导致循环
}

File: app/src/main/java/com/example/facecheck/ui/auth/LoginActivity.kt
package com.example.facecheck.ui.auth

import android.content.Context
import android.content.Intent
import android.os.Bundle
import android.widget.Toast
import androidx.activity.ComponentActivity
import androidx.activity.compose.setContent
import androidx.compose.foundation.Image
import androidx.compose.foundation.layout.*
import androidx.compose.material3.*
import androidx.compose.runtime.*
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.runtime.livedata.observeAsState
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.layout.ContentScale
import androidx.compose.ui.platform.LocalContext
import androidx.compose.ui.res.painterResource
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.text.input.PasswordVisualTransformation
import androidx.compose.ui.unit.dp
import androidx.compose.ui.unit.sp
import androidx.compose.ui.viewinterop.AndroidView
import androidx.lifecycle.ViewModelProvider
import com.airbnb.lottie.LottieAnimationView
import com.airbnb.lottie.LottieDrawable
import com.example.facecheck.MainActivity
import com.example.facecheck.R
import com.example.facecheck.ui.theme.FaceCheckTheme
import kotlinx.coroutines.delay

class LoginActivity : ComponentActivity() {

    private lateinit var loginViewModel: LoginViewModel

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)

        loginViewModel = ViewModelProvider(this)[LoginViewModel::class.java]

        val prefs = getSharedPreferences("user_prefs", MODE_PRIVATE)
        val role = prefs.getString("user_role", "")
        if (role == "teacher" || role == "student") {
            startActivity(Intent(this, MainActivity::class.java))
            finish()
            return
        }

        setContent {
            FaceCheckTheme {
                LoginScreen(loginViewModel)
            }
        }
    }
}

@Composable
fun LoginScreen(viewModel: LoginViewModel) {
    val context = LocalContext.current
    var username by remember { mutableStateOf("") }
    var password by remember { mutableStateOf("") }
    var isStudentLogin by remember { mutableStateOf(false) }
    var rememberMe by remember { mutableStateOf(false) }
    var loginSuccessHandled by remember { mutableStateOf(false) }
    var showSuccessAnimation by remember { mutableStateOf(false) }
    val uiState by viewModel.uiState.observeAsState()

    LaunchedEffect(uiState) {
        when (val state = uiState) {
            is LoginUiState.Success -> {
                if (loginSuccessHandled) return@LaunchedEffect
                loginSuccessHandled = true
                Toast.makeText(context, "登录成功！", Toast.LENGTH_SHORT).show()
                val prefsEditor = context.getSharedPreferences("user_prefs", Context.MODE_PRIVATE).edit()
                prefsEditor.putString("user_role", state.role)
                prefsEditor.putString("access_token", state.accessToken)
                prefsEditor.putString("refresh_token", state.refreshToken)
                if (state.role == "teacher") {
                    prefsEditor.putLong("teacher_id", state.userId)
                } else {
                    prefsEditor.putLong("student_id", state.userId)
                }
                if (rememberMe) {
                    prefsEditor.putString("saved_username", username)
                    prefsEditor.putString("saved_password", password)
                    prefsEditor.putBoolean("remember_me", true)
                } else {
                    prefsEditor.remove("saved_username")
                    prefsEditor.remove("saved_password")
                    prefsEditor.remove("remember_me")
                }
                prefsEditor.apply()
                showSuccessAnimation = true
            }
            is LoginUiState.Error -> {
                Toast.makeText(context, state.message, Toast.LENGTH_SHORT).show()
                loginSuccessHandled = false
            }
            else -> {}
        }
    }

    LaunchedEffect(showSuccessAnimation) {
        if (showSuccessAnimation) {
            delay(1400)
            context.startActivity(Intent(context, MainActivity::class.java).apply {
                flags = Intent.FLAG_ACTIVITY_NEW_TASK or Intent.FLAG_ACTIVITY_CLEAR_TASK
            })
        }
    }

    LaunchedEffect(Unit) {
        val prefs = context.getSharedPreferences("user_prefs", Context.MODE_PRIVATE)
        rememberMe = prefs.getBoolean("remember_me", false)
        if (rememberMe) {
            username = prefs.getString("saved_username", "") ?: ""
            password = prefs.getString("saved_password", "") ?: ""
        }
    }

    Surface(modifier = Modifier.fillMaxSize(), color = MaterialTheme.colorScheme.background) {
        Box(modifier = Modifier.fillMaxSize()) {
            Column(
                modifier = Modifier
                    .fillMaxSize()
                    .padding(horizontal = 32.dp),
                verticalArrangement = Arrangement.Center,
                horizontalAlignment = Alignment.CenterHorizontally
            ) {
            Box(modifier = Modifier.fillMaxWidth()) {
                Image(
                    painter = painterResource(id = R.drawable.logo_whitebackground),
                    contentDescription = "App Logo",
                    modifier = Modifier
                        .fillMaxWidth()
                        .height(80.dp),
                    contentScale = ContentScale.FillWidth
                )
            }
            Spacer(modifier = Modifier.height(16.dp))
            Text(
                text = "多模式签到系统",
                style = MaterialTheme.typography.titleMedium,
                color = MaterialTheme.colorScheme.onSurface.copy(alpha = 0.7f)
            )
            Spacer(modifier = Modifier.height(48.dp))

            OutlinedTextField(
                value = username,
                onValueChange = { username = it },
                label = { Text(if (isStudentLogin) "学号" else "用户名") },
                modifier = Modifier.fillMaxWidth()
            )
            Spacer(modifier = Modifier.height(16.dp))

            OutlinedTextField(
                value = password,
                onValueChange = { password = it },
                label = { Text("密码") },
                modifier = Modifier.fillMaxWidth(),
                visualTransformation = PasswordVisualTransformation()
            )
            Spacer(modifier = Modifier.height(16.dp))

            Row(
                modifier = Modifier.fillMaxWidth(),
                horizontalArrangement = Arrangement.SpaceBetween,
                verticalAlignment = Alignment.CenterVertically
            ) {
                Row(verticalAlignment = Alignment.CenterVertically) {
                    Checkbox(checked = rememberMe, onCheckedChange = { rememberMe = it })
                    Text("记住密码", fontSize = 14.sp)
                }

                Row(verticalAlignment = Alignment.CenterVertically) {
                    Text(
                        text = if (isStudentLogin) "学生登录" else "教师登录",
                        fontSize = 14.sp,
                        color = MaterialTheme.colorScheme.primary
                    )
                    Spacer(modifier = Modifier.width(8.dp))
                    Switch(
                        checked = isStudentLogin,
                        onCheckedChange = { isStudentLogin = it }
                    )
                }
            }
            Spacer(modifier = Modifier.height(24.dp))

            Button(
                onClick = { viewModel.login(username, password, isStudentLogin) },
                modifier = Modifier.fillMaxWidth().height(56.dp),
                enabled = uiState !is LoginUiState.Loading
            ) {
                if (uiState is LoginUiState.Loading) {
                    CircularProgressIndicator(color = MaterialTheme.colorScheme.onPrimary)
                } else {
                    Text("登录", fontSize = 16.sp, fontWeight = FontWeight.Bold)
                }
            }
            Spacer(modifier = Modifier.height(12.dp))

            TextButton(onClick = { context.startActivity(Intent(context, RegisterActivity::class.java)) }) {
                Text("还没有账号？立即注册")
            }
            }
            if (showSuccessAnimation) {
                Box(
                    modifier = Modifier
                        .fillMaxSize(),
                    contentAlignment = Alignment.Center
                ) {
                    Surface(
                        modifier = Modifier.fillMaxSize(),
                        color = Color.Black.copy(alpha = 0.45f)
                    ) {}
                    AndroidView(
                        factory = { ctx ->
                            LottieAnimationView(ctx).apply {
                                setAnimation("lottie/telegram.json")
                                repeatCount = 0
                                repeatMode = LottieDrawable.RESTART
                                playAnimation()
                            }
                        },
                        modifier = Modifier
                            .fillMaxWidth()
                            .height(260.dp)
                    )
                }
            }
        }
    }
}
